connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java

```diff
@@ -526,165 +421,174 @@ private List<WorkerLoad> pickCandidateWorkerForReassignment(List<WorkerLoad> com
-                resources.tasks().add(tasks.next());
+        final Map<String, Set<E>> result = new HashMap<>();
+        // How many workers we've allocated a single extra resource instance to
+        int allocatedExtras = 0;
+        for (WorkerLoad worker : workers) {
+            int currentAllocationSizeForWorker = workerAllocationSize.apply(worker);
+            if (currentAllocationSizeForWorker <= minAllocatedPerWorker) {
+                // This worker isn't allocated more than the minimum; no need to revoke anything
+                continue;
+            }
+            int maxAllocationForWorker;
+            if (allocatedExtras < extrasToAllocate) {
+                // We'll allocate one of the extra resource instances to this worker
+                allocatedExtras++;
+                if (currentAllocationSizeForWorker == minAllocatedPerWorker + 1) {
+                    // If the worker's running exactly one more than the minimum, and we're allowed to
+                    // allocate an extra to it, there's no need to revoke anything
+                    continue;
+                }
+                maxAllocationForWorker = minAllocatedPerWorker + 1;
+            } else {
+                maxAllocationForWorker = minAllocatedPerWorker;
+            }
+
+            Set<E> revokedFromWorker = new LinkedHashSet<>();
+            result.put(worker.worker(), revokedFromWorker);
+
+            Iterator<E> currentWorkerAllocation = workerAllocation.apply(worker).iterator();
+            // Revoke resources from the worker until it isn't allocated any more than it should be
+            while (workerAllocationSize.apply(worker) > maxAllocationForWorker) {
+                if (!currentWorkerAllocation.hasNext()) {
+                    // Should never happen, but better to log a warning and move on than die and fail the whole rebalance if it does
+                    log.warn(
+                            "Unexpectedly ran out of {}s to revoke from worker {} while performing load-balancing revocations; " +
+                                    "worker appears to still be allocated {} instances, which is more than the intended allocation of {}",
+                            allocatedResourceName,
+                            worker.worker(),
+                            workerAllocationSize.apply(worker),
+                            maxAllocationForWorker
+                    );
+                    break;
+                }
+                E revocation = currentWorkerAllocation.next();
+                revokedFromWorker.add(revocation);
+                // Make sure to remove the resource from the worker load so that later operations
+                // (such as assigning newly-created connectors and tasks) can take that into account
+                currentWorkerAllocation.remove();
             }
         }
+        return result;
+    }
 
-        return revoking;
+    private int calculateDelay(long now) {
```

Comment:
I could see using many final parameters in Kafka code. So this just for code convention and safety.

```suggestion
    private int calculateDelay(final long now) {
```